compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmInstructionGen.java

```diff
@@ -1520,17 +1521,63 @@ void generateCastIns(BIRNonTerminator.TypeCast typeCastIns) {
     }
 
     void generateTypeTestIns(BIRNonTerminator.TypeTest typeTestIns) {
-        // load source value
-        this.loadVar(typeTestIns.rhsOp.variableDcl);
-
-        // load targetType
-        jvmTypeGen.loadType(this.mv, typeTestIns.type);
+        var sourceValue = typeTestIns.rhsOp.variableDcl;
+        BType targetType = typeTestIns.type;
+        if (canOptimizeType(sourceValue, targetType)) {
+            handleErrorUnionType(typeTestIns);
+            return;
+        }
+        this.loadVar(sourceValue);
+        jvmTypeGen.loadType(this.mv, targetType);
 
         this.mv.visitMethodInsn(INVOKESTATIC, TYPE_CHECKER, "checkIsType",
                                 String.format("(L%s;L%s;)Z", OBJECT, TYPE), false);
         this.storeToVar(typeTestIns.lhsOp.variableDcl);
     }
 
+    private void handleErrorUnionType(BIRNonTerminator.TypeTest typeTestIns) {
+        loadVar(typeTestIns.rhsOp.variableDcl);
+        mv.visitTypeInsn(INSTANCEOF, BERROR);
+        if (typeTestIns.type.tag != TypeTags.ERROR) {
+            generateNegateBoolean();
+        }
+        storeToVar(typeTestIns.lhsOp.variableDcl);
+    }
+
+    private boolean canOptimizeType(BIRNode.BIRVariableDcl rhsVar, BType type) {
+        BType rhsType = rhsVar.type;
+        if (rhsType.tag != TypeTags.UNION) {
+            return false;
+        }
+        var unionMemberTypes = ((BUnionType) rhsType).getMemberTypes();
+        if (unionMemberTypes.size() != 2) {
+            return false;
+        }
+        BType errorType = null;
+        BType otherType = null;
+        boolean foundError = false;
+        for (BType bType : unionMemberTypes) {
+            if (bType.tag == TypeTags.ERROR) {
+                errorType = bType;
+                foundError = true;
+            } else {
+                otherType = bType;
+            }
+        }
+        return foundError && (type.equals(errorType) || type.equals(otherType));
```

Comment:
hmm this looks like insufficient checking. Does this work if it's a union of `MyError1|MyError2` case?

Reply:
Thanks that would cause an issue. Fixed here:  https://github.com/ballerina-platform/ballerina-lang/pull/31228